data_handler/handlers/loan_states/hashtack_v0/interest_rate.py

```diff
@@ -1,87 +1,168 @@
+            supply_apr_bps = Decimal(int(event["data"][2], 16))
 
             # Calculate interest rate for supply and borrow and convert to percents using the formula:
             # (apr * seconds_passed / seconds_in_year) / 10000
-            supply_interest_rate_cum += (supply_apr * seconds_passed / SECONDS_IN_YEAR) / percents_decimals_shift
-            borrow_interest_rate_cum += (borrow_apr * seconds_passed / SECONDS_IN_YEAR) / percents_decimals_shift
-            current_event = HashstackV0Event(
-                event["block_number"], current_timestamp, supply_interest_rate_cum, borrow_interest_rate_cum,
+            current_collateral_change = (supply_apr_bps * seconds_passed / SECONDS_IN_YEAR) * percents_decimals_shift
+            current_debt_change = (borrow_apr_bps * seconds_passed / SECONDS_IN_YEAR) * percents_decimals_shift
+            interest_rate_state.update_state_cumulative_data(
+                token_name, event["block_number"], current_collateral_change, current_debt_change
             )
-            self.interest_rates.append(current_event)
 
-    def serialize(self) -> list[InterestRate]:
-        """Get interests rates serialized as database models."""
-        # TODO: Add integration with existing sqlalchemy model
-        return [event.serialize() for event in self.interest_rates]
+        # Write last block data
+        self._add_block_data(blocks_data, interest_rate_state.build_interest_rate_model(HASHSTACK_ID))
+        return blocks_data
+
+    def _get_blocks_bounds(self, start_block, end_block, latest_block) -> tuple[int, int]:
+        """
+        Calculate the bounds for the blocks pagination.
+        :param start_block: int - Previous start block number.
+        :param end_block: int - Previous end block number.
+        :param latest_block: int - The latest block number.
+        :return: tuple[int, int] - The new start and end block numbers.
+        """
+        if end_block + self.PAGINATION_SIZE < latest_block:
+            start_block += end_block
+        else:
+            end_block = latest_block
+            return start_block, end_block
+        if start_block + self.PAGINATION_SIZE <= latest_block:
+            end_block += self.PAGINATION_SIZE
+        else:
+            end_block = latest_block
+        return start_block, end_block
+
+    async def _run_async(self) -> None:
+        """Asynchronous function for running the interest rate calculation process and fetching data from on-chain."""
+        latest_block = await NET.get_block_number()
+        start_block = self.last_block_data.block if self.last_block_data else 0
+        if start_block == latest_block:
+            return
+        start_block, end_block = self._get_blocks_bounds(
+            start_block,
+            start_block + self.PAGINATION_SIZE,
+            latest_block
+        )
+        events_over = False
+        # Fetch and set events until blocks are over
+        while not events_over:
```

Comment:
The iteration over chunks of size 1000 seems a little complicated. Wouldn't this do the job?
```suggestion
        while start_block < latest_block:
                start_block += self.PAGINATION_SIZE
                end_block = start_block + self.PAGINATION_SIZE
```

Reply:
Pls consider this one more time. The `_get_blocks_bounds` method seems quite redundant.